### PR TITLE
Upload logs to Warehouse via SFTP (no-remove)

### PR DESCRIPTION
**Pull request by rgaudin**
## Rationale

#174 we want to keep a copy of the full log of the main container

## Changes

Copy of #238 without the code required to remove the previous log (api endpoint and test, api access from worker and file removal in worker).

### Warehouse

* changed root to `/files/` instead of `/zim_files` (for semantic reason)
* kept zim uploads in `/files/zim` but creates a `/files/logs` folder as well to receive logs.
* added support for the `remove()` command (we'll need that soon to remove logs when task's EOL is implemented)

### Worker

* get full output log (`stdout` + `stderr`) from docker
* save it to a temporary file while compressing it into `xz`: compression is done in batch of 1M lines
* upload the log to `/logs/<container_name>.log.xz`. Chose this so we could one day have multiple logs per task (redis for instance)
* upload log info saved to `ContainerResult`
* slight refactoring of upload code (for files)

### monitor

* on container finish event, store log file info into mogo (filename and size)

### U.I

* added a link in task detail to the compressed log file (via HTTP/s, already deployed on server)
* rebuilt angular project

### CODE REVIEW

File: warehouse/app/warehouse/__init__.py [hunk 1]
```diff
@@ -74,13 +74,15 @@ def _configure_root_dir():
...
-                os.makedirs(root)
-        else:
-            os.makedirs(root)
+        root = os.getenv('ROOT_PATH', '/files')
+
+        for subdir_name in ('zim', 'logs'):
+            subdir = os.path.join(root, subdir_name)
+            if os.path.exists(subdir):
+                if not os.path.isdir(subdir):
+                    os.remove(subdir)
```
Comment (automactic) [inline warehouse/app/warehouse/__init__.py]: I am not sure, but if the subdir has content in it, is `remove` gonna be successful?

Comment (automactic) [reply warehouse/app/warehouse/__init__.py]: According to the [docs](https://docs.python.org/3.7/library/os.html#os.remove), `rmdir` is needed to remove a dir

File: warehouse/app/warehouse/sftp.py [hunk 2]
```diff
@@ -157,7 +157,23 @@ def readlink(self, path):
...
+        # check permission
+        if not path.startswith(self.root):
+            return sftp.SFTP_PERMISSION_DENIED
+
+        # return error if file doesn't exists
+        if not os.path.exists(path):
+            return sftp.SFTP_FAILURE
+
+        try:
+            os.unlink(path)
```
Comment (automactic) [inline warehouse/app/warehouse/sftp.py]: should we consider the case if path is a dir?

File: worker/app/operations/upload.py [hunk 3]
```diff
@@ -30,9 +40,52 @@ def execute(self):
...
                         continue
                     client.upload_file(self.remote_working_dir, file)
+
+                    if self.delete:
+                        file.unlink()
         except SSHException as e:
             raise UploadError(code='paramiko.Error', message=str(e))
+
+    @staticmethod
+    def upload(remote_working_dir: str, files: list = None, directory: Path = None, delete: bool = False):
```
Comment (automactic) [inline worker/app/operations/upload.py]: Personally I am not a fan of having `files` and `directory` and ask user to provide one of them. What happens if I provide both, do they / should they all get uploaded? 

Maybe we should have `upload_files` and `upload_directory_content`, one could call the other? What do you think?

File: worker/app/operations/upload.py [hunk 4]
```diff
@@ -30,9 +40,52 @@ def execute(self):
...
+
+        if files is None and directory is None:
+            raise ValueError("upload requires either `files` or `directory`")
+
+        # build list of files if invoked with directory
+        if files is None:
+            files = [file for file in directory.iterdir() if not file.is_dir()]
+
+        stats = [{'name': file.name, 'size': file.stat().st_size} for file in files]
+        files_desc = ','.join(['{name} - {size}'.format(**stat) for stat in stats])
```
Comment (automactic) [inline worker/app/operations/upload.py]: nit: would you consider consolidate these 2 line into one line with python3.7 style string interpolation?

File: worker/app/operations/upload.py [hunk 5]
```diff
@@ -30,9 +40,52 @@ def execute(self):
...
+
+        upload = Upload(remote_working_dir=remote_working_dir, files=files)
+        upload.execute()
+
+        return stats
+
+    @staticmethod
+    def upload_log(container):
+        """fetch, compress and upload the container's stdout to the warehouse"""
+        container.reload()
```
Comment (automactic) [inline worker/app/operations/upload.py]: Why do we need `reload`? 🤔

File: worker/app/tasks/base.py [hunk 6]
```diff
@@ -52,22 +53,13 @@ def clean_up(self):
...
+        shutil.rmtree(working_dir, ignore_errors=True)
 
     def on_success(self, retval, task_id, args, kwargs):
         self.clean_up()
 
     def on_failure(self, exc, task_id, args, kwargs, einfo):
         self.clean_up()
+
+    def upload_zims(self, remote_working_dir: str, files: list = None, directory: Path = None):
+        return Upload.upload(f'/zim{remote_working_dir}', files, directory)
```
Comment (automactic) [inline worker/app/tasks/base.py]: Hmmm if this is just one line, I am not sure it is really beneficial to make it a separate function.

File: dispatcher/frontend/src/app/task/task-detail/task-detail.ts [hunk 7]
```diff
@@ -4,12 +4,16 @@ import { Observable } from 'rxjs';
 import { switchMap } from 'rxjs/operators';
 
 import { Task, TasksService } from '../../services/task.service';
+import { getWareHouseLogsUrl } from '../../services/config';
```
[the remark this answers is not in the document]
Comment (rgaudin) [reply dispatcher/frontend/src/app/task/task-detail/task-detail.ts]: Ah indeed you introduced that after. I'll merge master-in and use the base service

File: worker/app/operations/upload.py (hunk 3 above)
Comment (rgaudin) [reply worker/app/operations/upload.py]: 👍

File: worker/app/operations/upload.py (hunk 5 above)
Comment (rgaudin) [reply worker/app/operations/upload.py]: you're right, we don't. `logs()` is not using cached data.

File: worker/app/tasks/base.py (hunk 6 above)
Comment (rgaudin) [reply worker/app/tasks/base.py]: I'd prefer to keep it even though it's simple as it's gonna be used in every scraper and makes sure we keep that `/zim` prefix in place.

File: warehouse/app/warehouse/sftp.py (hunk 2 above)
Comment (rgaudin) [reply warehouse/app/warehouse/sftp.py]: right.

File: worker/app/operations/upload.py (hunk 4 above)
Comment (rgaudin) [reply worker/app/operations/upload.py]: not sure what you mean. I switched to literal string for `files_desc` but don't want to combined them as `stats` is useful and returned while `files_desc` is just for logging.

**Pull request merged**